index.html

```diff
@@ -232,39 +235,393 @@ <h3>Create</h3>
+the resulting output.
+          </p>
+
+          <ol class="algorithm">
+            <li>
+Initialize <em>verification method</em> to an empty object.
+            </li>
+            <li>
+Set <em>multicodec value</em> and <em>raw public key bytes</em> to the result
+of passing <em>multibase value</em> and <em>options</em> to
+<a href="#derive-encryption-key-algorithm"></a>.
+            </li>
+            <li>
+Ensure the proper key length of <em>raw public key bytes</em> based on the
+<em>multicodec value</em> table provided below:
+              <table class="simple">
+                <thead>
+                  <th>Multicodec hexadecimal value</th>
+                  <th>public key byte length</th>
+                  <th>Description</th>
+                </thead>
+                <tbody>
+                  <tr>
+                    <td>`0xec`</td>
+                    <td>32 bytes</td>
+                    <td>x25519-pub - Curve25519 public key</td>
+                  </tr>
+                </tbody>
+              </table>
+If the byte length of <em>raw public key bytes</em> does not match the expected
+public key length for the associated <em>multicodec value</em>, an
+INVALID_PUBLIC_KEY_LENGTH error MUST be raised.
+            </li>
+            <li>
+Create the <em>multibase value</em> by concatenating the letter 'z' and the
+base58-btc encoding of the concatenation of the <em>multicodec value</em> and
+the <em>raw public key bytes</em>.
+            </li>
+            <li>
+Set the <em>verification method</em> `id` value by concatenating
+<em>identifier</em>, a hash character (`#`), and the <em>multibase value</em>.
+            </li>
+            <li>
+Set the <em>key encoding type</em> value to the <em>options</em> `type` value
+if it is set, or `Multikey` if it is not set.
+            </li>
+            <li>
+If <em>options</em> `allowExperimentalPublicKeyTypes` is not set to `true` and
+<em>key encoding type</em> is not `Multikey`, `JsonWebKey2020`, or
+`X25519KeyAgreementKey2020`, an INVALID_PUBLIC_KEY_TYPE error MUST be raised.
+            </li>
+            <li>
+Set the <em>verification method</em> `type` to the <em>key encoding type</em>.
+            </li>
+            <li>
+Set the <em>verification method</em> `controller` to <em>identifier</em>.
+            </li>
+            <li>
+If <em>key encoding type</em> is `Multikey` or `X25519KeyAgreementKey2020`,
+set the <em>verification method</em> `publicKeyMultibase` value to
+<em>multibase value</em>.
+            <li>
+If <em>key encoding type</em> is `JsonWebKey2020`, set the
+<em>verification method</em> `publicKeyJwk` value to the result of
+passing <em>multicodec value</em> and <em>raw public key bytes</em> to
+<a href="#encode-jwk-algorithm"></a>.
+            </li>
+            <li>
+Return <em>verification method</em>.
+            </li>
+          </ol>
+
+        </section>
+
+        <section class="normative">
+          <h4>Derive Encryption Key Algorithm</h4>
```

Comment:
its no clear to me which `option` this applies to.

I would expect to be able to do the following:

```
data0 = did.key.create({ ed26619 } )
data1 = did.key.create({ curve25519 } )
data2= did.key.create({ ed26619, curve25519 } )
```

I think this PR does not address the ambiguity of which of these representations you get, unless you are saying derive always happens if its possible... in which case 0 and 2 and the same.

Reply:
> its not clear to me which option this applies to.

See steps 5 and 9 in Section 3.1.1 Document Creation Algorithm, namely this text:

> 5) Initialize the encryption verification method to the result of passing identifier, multibase value, and options to [§ 3.1.5 Encryption Method Creation Algorithm](https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#encryption-method-creation-algorithm).
> ...
> 9) If options disableKeyAgreement is not set to true, add the encryption verification method value to the verificationMethod array and initialize the keyAgreement property in document to an array where the first item is the value of the id property in encryption verification method.

https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#document-creation-algorithm

... which then involves step 2 in Section 3.1.5 Encryption Method Creation Algorithm, namely this text:

> Set multicodec value and raw public key bytes to the result of passing multibase value and options to [§ 3.1.6 Derive Encryption Key Algorithm](https://pr-preview.s3.amazonaws.com/w3c-ccg/did-method-key/pull/51.html#derive-encryption-key-algorithm).

So, at present, derive always happens, but keyAgreement isn't set if `disableKeyAgreement` is set to true.

We should probably change this such that key derivation doesn't happen if `disableKeyAgreement` is set to true.

I'll try to rework that language to be a bit more clear about setting the keyAgreement value. Would that work for you? What would you prefer?

Reply:
yeah, lets split this up, there are 2 things happening (our libraries already allow you to ask for variations of these for NIST curves / X25519).

1. Do you want key agreement?
2. Do you want key derivation?

If you ask for key agreement but don't allow for derivation for ed25519 you should get an error.

If you ask for derivation with P256, you should get an error.

Point conversion should not be a side effect of asking for key agreement, it should be an explicit request, or an explicit default.

Reply:
> Point conversion should not be a side effect of asking for key agreement, it should be an explicit request, or an explicit default.

But isn't the only way you get key agreement w/ did:key via key derivation? Why would you pick `allowKeyDerivation=true` but then do `addKeyAgreement=true`?

Reply:
Took a cut at it here by adding a new `enableEncryptionKeyDerivation` flag and switching whether `keyAgreement` is set off of that: cc09b8eb45f97192fe79dc3bc44bbbbfed592fe1

Reply:
> But isn't the only way you get key agreement w/ did:key via key derivation?

No, NIST Curves support keyAgreement with the same key they use to support signatures.

Reply:
`enableEncryptionKeyDerivation` I suppose you could use this to turn keyAgreement on or off for P-256?

Reply:
> No, NIST Curves support keyAgreement with the same key they use to support signatures.

Sure, but that would never pass a FIPS audit -- it violates the "don't use the same key for signing and KEM" guidance -- and it's especially bad for P-256, that knowingly weakens the shared key material.

> enableEncryptionKeyDerivation I suppose you could use this to turn keyAgreement on or off for P-256?

Yeah, but I guess that would probably be a misuse of that particular flag.

Perhaps we need an `onlyGenerateKeyAgreement` flag for P-256, so you can generate a did:key document for signing OR you can generate a did:key document only for `keyAgreement`, but you should definitely not generate one for both.

Another possibility is to create a variation of did:key that encodes the signing public key and the encryption key in the did:key itself, perhaps separated by a colon, so a P-256 did:key that supports both signing and encryption could look like this:

`did:key:zDnaerDaTF5BXEavCrfRZEk316dpbLsfPDZ3WJ5hRTPFU2169:zDnaerx9CtbPJ1q36T5Ln5wYt3MQYeGRG5ehnPAmxcf5mDZpv`

... but that's sorta gross.

We should probably raise an issue for this. I didn't think anyone was trying to support both signing and encryption with did:keys and any non-Montgomery coordinate system.

Reply:
We should explicitly disallow signing and encryption from a single did:key

Reply:
> We should explicitly disallow signing and encryption from a single did:key

Except where there are IETF documents by CFRG folks with mathematical proofs showing that it's safe to do so, such as with ECDH w/ Montgomery Curve public key derivation (OSCORE, THORMARKER), but with an option to turn that feature off for those that need to demonstrate FIPS compliance, right? :)

Reply:
I'm not sure point conversion will pass a FIPS audit either, best not to require it or default to it.... instead allow for it to be configured.

If you want to use the same did:key for signing and encryption you can. 

If you want to get a presentation of a did key that only supports signing (or only supports encryption) you should be able to do this... BTW, a better solution is to just use the x2559 key as its own separate did... like this:

https://did.key.transmute.industries/generate/x25519?seed=d197db5d456b1be622144c8f0b6b12602fe93b29e65d8e47a15d67abe8513fbc

https://did.key.transmute.industries/generate/ed25519?seed=49f85b0fdce80cfed8db5daad14e42d4d9493add3b95281dc56a60ebd2659d5d

This allows you to: use signing and encryption with did key, not reuse entropy, not do point conversion... this is likely to be looked on more favorably during an audit.

Reply:
todays's did key implementations don't allow you to isolate the key purpose, and the ed25519 variant actually encourages entropy reuse under the assumption that point conversion is safe... this is essentially treating experimental behavior as a safe default... and still violating the "use a key for a single purpose" NIST recommendation.

Reply:
The nature of the did key spec is that its an identifer for a public key... not an identifer for nist recommendations of a public key and its use... better to use did:jwk for that, and set the `use` property before encoding... https://github.com/quartzjer/did-jwk/blob/main/spec.md

Reply:
@OR13,

> The ed25519 variant actually encourages entropy reuse under the assumption that point conversion is safe...

I recommend that we start saying "...based on a [proof](https://eprint.iacr.org/2021/509) that point conversion is safe..." instead of "under the assumption" to avoid misleading people that there is no such proof and it is a bare assumption only.

Reply:
I should also add -- expressing the same key in multiple ways is not a violation -- and that's all that we're talking about here.  Whether or not that key is *used* is a separate matter in my view -- and unrelated to what this spec is doing. This spec does not say "you MUST use this key" nor would that be enforceable.